compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -483,7 +484,8 @@ object desugar {
         val enumCompanionRef = TermRefTree()
         val enumImport =
           Import(enumCompanionRef, enumCases.flatMap(caseIds).map(ImportSelector(_)))
-        (enumImport :: enumStats, enumCases, enumCompanionRef)
+        val enumSpecMethods = EnumGetters()
```

Comment:
An alternative to a specialised `EnumGetters` tree could be a `FlatTree` with defs for `ordinal` and `enumLabel` where the `ordinal` method is filtered out when the parent is `java.lang.Enum`, but this is more lazy

Reply:
What about adding those defs when typing the enum itself where presumably the parents are known?

Reply:
I'd prefer to desugar this into a concrete, synthetic ordinal method with an empty body. Then we do not need any special treatment in Namer, and Typer can fill in the body. I believe this would be simpler than creating a new kind of tree (EnumGetters) that has to be treated specially in various places.

Reply:
This one I am more confused about. If the enum extends `java.lang.Enum` then we can not declare an override for `ordinal` as it is final. Is there a way to define it with a tree without causing any follow on errors? e.g. the child cases of the enum then have to declare an implementation for the now abstract `ordinal` which will be forbidden

Reply:
never mind, i see the special case in typedDefDef and I can adapt it to work

Reply:
I have updated the implementation to use the same mechanism as clashing case class synthetic methods